### ISSUE TITLE
GRPO questions

### ISSUE THREAD

**Issue by natolambert** [huggingface/trl#2608]
Hey friends! I have some questions on the GRPO implementation, happy to discuss.

1. It looks like you apply the KL distance in the advantages, while the DeepSeekMath paper says “Also note that, instead of adding KLpenalty in the reward, GRPO regularizes by directly adding the KL divergence between the trained policy and the reference policy to the loss, avoiding complicating the calculation of 𝐴ˆ”
2. Did any thought go into making this a sum of loss and not mean? We aren’t sure https://github.com/huggingface/trl/blob/fe4b5efe4e23f4331ba9c5b0c8bd92dc8302c287/trl/trainer/grpo_trainer.py#L286
3. I didn’t see the PPO clipping logic in policy gradient loss, coming soon?

**Comment by qgallouedec** [comment]
Hey @natolambert!!


> It looks like you apply the KL distance in the advantages, while the DeepSeekMath paper says “Also note that, instead of adding KLpenalty in the reward, GRPO regularizes by directly adding the KL divergence between the trained policy and the reference policy to the loss, avoiding complicating the calculation of 𝐴ˆ”

Here's how I understand it:
In PPO, the KL div is subtracted from the reward (_adding KL penalty in the reward_)

https://github.com/huggingface/trl/blob/949db2357e62d2f0a34decfc5e87eeeea0c6d72c/trl/trainer/ppo_trainer.py#L509-L515

Here, in GRPO, the advantage is computed without the KL term. It's just the output of the reward function, normalised per group:

https://github.com/huggingface/trl/blob/949db2357e62d2f0a34decfc5e87eeeea0c6d72c/trl/trainer/grpo_trainer.py#L274-L281

and later you subtract the KL term

https://github.com/huggingface/trl/blob/949db2357e62d2f0a34decfc5e87eeeea0c6d72c/trl/trainer/grpo_trainer.py#L285

I don't know how complicated it would have been to integrate KL into the reward. It would probably look something like

```python
# Subtract KL
rewards  = rewards - self.beta * per_token_kl  # <- THIS IS ADDED

# Compute grouped-wise rewards
mean_grouped_rewards = rewards.view(-1, self.num_generations).mean(dim=1)
std_grouped_rewards = rewards.view(-1, self.num_generations).std(dim=1)

# Normalize the rewards to compute the advantages
mean_grouped_rewards = mean_grouped_rewards.repeat_interleave(self.num_generations, dim=0)
std_grouped_rewards = std_grouped_rewards.repeat_interleave(self.num_generations, dim=0)
advantages = (rewards - mean_grouped_rewards) / (std_grouped_rewards + 1e-4)

# x - x.detach() allows for preserving gradients from x
advantages = torch.exp(per_token_logps - per_token_logps.detach()) * advantages.unsqueeze(1)
per_token_loss = -advantages  # <- THIS IS MODIFIED
```


which seems pretty simple. But perhaps he means that the subsequent calculations would have been more complicated.

**Comment by qgallouedec** [comment]
> 3. I didn’t see the PPO clipping logic in policy gradient loss, coming soon?

In the current implementation, we're just update once after ageneration. In fact, we align with this sentence from the paper:

> The policy model only has a single update following each exploration stage.


Therefore It implies that $\pi_{\theta_{\text{old}}} = \pi_\theta$, and the equation

```math
\mathcal{J}_{\text{GRPO}}(\theta) =\frac{1}{G} \sum_{i=1}^G \frac{1}{|o_i|}\sum_{t=1}^{|o_i|}\left[\min \left(\frac{\pi_\theta(o_{i,t} | q, o_{i,< t})}{\pi_{\theta_{\text{old}}}(o_{i,t} | q, o_{i,< t})} \hat{A}_{i,t}, \text{clip}\left(\frac{\pi_\theta(o_{i,t} | q, o_{i,< t})}{\pi_{\theta_{\text{old}}}(o_{i,t} | q, o_{i,< t})}, 1 - \epsilon, 1 + \epsilon\right) \hat{A}_{i,t}\right) - \beta \mathbb{D}_{\text{KL}}\left[\pi_\theta \| \pi_{\text{ref}}\right]\right]
```
 
can be simplified to

```math
\mathcal{J}_{\text{GRPO}}(\theta) = \frac{1}{G} \sum_{i=1}^G \frac{1}{|o_i|} \sum_{t=1}^{|o_i|} \left[ \frac{\pi_\theta(o_{i,t} | q, o_{i,< t})}{\left[\pi_\theta(o_{i,t} | q, o_{i,< t})\right]_\cancel{\nabla}} \hat{A}_{i,t} - \beta \mathbb{D}_{\text{KL}}\left[\pi_\theta \| \pi_{\text{ref}}\right] \right].
```

**But**, we could support multiple update after each generation. And it would require to have this PPO clipping logic. It would probably allow to reuse generation and be more sample efficient. On the other hand, this would probably require a rather hard-to-read implementation, as the optimization step is performed in the parent trainer class. It would look something like

```python
def __init__(self, ...):
    ...
    self.train_dataset = repeat_interleave(train_dataset, self.num_grpo_iterations)  # [prompt0, prompt1] -> [prompt0, prompt0, prompt0, prompt1, prompt1, prompt1]

def compute_loss(self, model, inputs):
    if self.step % self.num_grpo_iterations == 0: # self.num_grpo_iterations  is 𝜇 in the paper
        completions = model.generate(prompts)
        self.old_log_probs = model(cat(prompts, completions))

    log_probs = model(cat(prompts, completions))
    log_ratio = log_probs - self.old_log_probs
    losses = min(exp(log_ratio)*advantages, clip(exp(log_ratio), 1-epsilon, 1+epsilon)*advantages)
    losses = losses - beta*kl
```

not sure if it's worth having this extra complexity.

**Comment by SeunghyunSEO** [comment]
> Hey @natolambert!!
> 
> 
> > 1. It looks like you apply the KL distance in the advantages, while the DeepSeekMath paper says “Also note that, instead of adding KLpenalty in the reward, GRPO regularizes by directly adding the KL divergence between the trained policy and the reference policy to the loss, avoiding complicating the calculation of 𝐴ˆ”
> 
> Here's how I understand it:
> In PPO, the KL div is subtracted from the reward (_adding KL penalty in the reward_)
> 
> https://github.com/huggingface/trl/blob/949db2357e62d2f0a34decfc5e87eeeea0c6d72c/trl/trainer/ppo_trainer.py#L509-L515
> 
> Here, in GRPO, the advantage is computed without the KL term. It's just the output of the reward function, normalised per group:
> 
> https://github.com/huggingface/trl/blob/949db2357e62d2f0a34decfc5e87eeeea0c6d72c/trl/trainer/grpo_trainer.py#L274-L281
> 
> and later you subtract the KL term
> 
> https://github.com/huggingface/trl/blob/949db2357e62d2f0a34decfc5e87eeeea0c6d72c/trl/trainer/grpo_trainer.py#L285
> 
> I don't know how complicated it would have been to integrate KL into the reward. It would probably look something like
> 
>
> # Subtract KL
> rewards  = rewards - self.beta * per_token_kl  # <- THIS IS ADDED
> 
> # Compute grouped-wise rewards
> mean_grouped_rewards = rewards.view(-1, self.num_generations).mean(dim=1)
> std_grouped_rewards = rewards.view(-1, self.num_generations).std(dim=1)
> 
> # Normalize the rewards to compute the advantages
> mean_grouped_rewards = mean_grouped_rewards.repeat_interleave(self.num_generations, dim=0)
> std_grouped_rewards = std_grouped_rewards.repeat_interleave(self.num_generations, dim=0)
> advantages = (rewards - mean_grouped_rewards) / (std_grouped_rewards + 1e-4)
> 
> # x - x.detach() allows for preserving gradients from x
> advantages = torch.exp(per_token_logps - per_token_logps.detach()) * advantages.unsqueeze(1)
> per_token_loss = -advantages  # <- THIS IS MODIFIED
> ```
> 
> 
> which seems pretty simple. But perhaps they means that the subsequent _equations and calculations_ would have been more complicated.

i think there is miscommunication.
@natolambert pointed out you should not abstract kl term in advantages term following original GRPO impl and trl impl follows paper well. but the variable name make readers confused.

```python
# Normalize the rewards to compute the advantages
        mean_grouped_rewards = mean_grouped_rewards.repeat_interleave(self.num_generations, dim=0)
        std_grouped_rewards = std_grouped_rewards.repeat_interleave(self.num_generations, dim=0)
        advantages = (rewards - mean_grouped_rewards) / (std_grouped_rewards + 1e-4)

        # x - x.detach() allows for preserving gradients from x
        advantages = torch.exp(per_token_logps - per_token_logps.detach()) * advantages.unsqueeze(1)
        per_token_loss = -(advantages - self.beta * per_token_kl)
        loss = ((per_token_loss * completion_mask).sum(dim=1) / completion_mask.sum(dim=1)).mean()
```

here `advantages` in last line means `per_token_prob * A - kld`.
and i think we should mean per group loss and sum per token loss following the paper.

![image](https://github.com/user-attachments/assets/84f3a4cd-2fe4-49c7-bdfd-fc52594a2989)

**Comment by qgallouedec** [comment]
So this instead (just a renaming)?:

```python
# x - x.detach() allows for preserving gradients from x
per_token_loss = torch.exp(per_token_logps - per_token_logps.detach()) * advantages.unsqueeze(1)
per_token_loss = -(per_token_loss - self.beta * per_token_kl)
loss = ((per_token_loss * completion_mask).sum(dim=1) / completion_mask.sum(dim=1)).mean()
```

**Comment by SeunghyunSEO** [comment]
> So this instead (just a renaming)?:
> 
> ```python
> # x - x.detach() allows for preserving gradients from x
> per_token_loss = torch.exp(per_token_logps - per_token_logps.detach()) * advantages.unsqueeze(1)
> per_token_loss = -(per_token_loss - self.beta * per_token_kl)
> loss = ((per_token_loss * completion_mask).sum(dim=1) / completion_mask.sum(dim=1)).mean()
> ```
> 
> Maybe more aligned with the formulation
> 
> > directly adding the KL divergence between the trained policy and the reference policy to the loss

lgtm :)

**Comment by qgallouedec** [comment]
#2616

**Comment by natolambert** [comment]
Yeah, I'm mostly aligned now (I haven't fully checked the math), but it seems like practically this is because you don't do minimatches you can get away with it? Do you have a line by line derivation (lol, I will ask cluade). 👀 

Thanks @SeunghyunSEO for the snipe on the variable names, that's what caught me up. Looks fine now and thanks for updating it.

**Comment by natolambert** [comment]
For example, you can see our implementation in open instruct that we just added too https://github.com/allenai/open-instruct/pull/523

**Comment by SonuDixit** [comment]
@qgallouedec 
> # x - x.detach() allows for preserving gradients from x
per_token_loss = torch.exp(per_token_logps - per_token_logps.detach()) * advantages.unsqueeze(1)

Here, If we take the gradient of per_token_loss with respect to per_token_logps , we get 0.  

Should this line be - 
per_token_loss = ( torch.exp(per_token_logps) - torch.exp(per_token_logps.detach()) ) * advantages.unsqueeze(1)

**Comment by ghrua** [comment]
> The policy model only has a single update following each exploration stage.

Hey @qgallouedec, thanks you very much for explaining the following code 

```
advantages = torch.exp(per_token_logps - per_token_logps.detach()) * advantages.unsqueeze(1)
```

But why not make the code clearer and simpler as follows:

```
advantages = per_token_logps * advantages.unsqueeze(1)
```

I think the two lines of code are equivalent, because the derivative of `exp(x)` is `exp(x) * d(x)`, where `exp(x) = exp(0) = 1` and `d(x) = d(per_token_logps)`.  In other words, the single-update GRPO degrades to the standard policy gradient, which is consistent with the Eq. (17) in DeepSeekMath. 

Please let me know if I misunderstood anything!

**Comment by ShinnosukeUesaka** [comment]
I think KL loss has not effect on the gradient because the ref model and model are identical in the current implementation

**Comment by qgallouedec** [comment]
Before the first optimization step yes. But after this the kl term has an impact
